engine/pkg/externalresource/manager/gc_coordinator.go

```diff
@@ -0,0 +1,249 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package manager
+
+import (
+	"context"
+	gerrors "errors"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tiflow/dm/pkg/log"
+	"go.uber.org/ratelimit"
+	"go.uber.org/zap"
+
+	libModel "github.com/pingcap/tiflow/engine/lib/model"
+	"github.com/pingcap/tiflow/engine/model"
+	resModel "github.com/pingcap/tiflow/engine/pkg/externalresource/resourcemeta/model"
+	"github.com/pingcap/tiflow/engine/pkg/notifier"
+	pkgOrm "github.com/pingcap/tiflow/engine/pkg/orm"
+)
+
+// DefaultGCCoordinator implements interface GCCoordinator.
+// It is responsible for triggering file resource garbage collection.
+type DefaultGCCoordinator struct {
+	executorInfos ExecutorInfoProvider
+	jobInfos      JobStatusProvider
+	metaClient    pkgOrm.ResourceClient
+	notifier      GCNotifier
+}
+
+// GCNotifier exposes a GCNotify method which is called
+// when a new object is marked as needing GC.
+type GCNotifier interface {
+	GCNotify()
+}
+
+// NewGCCoordinator creates a new DefaultGCCoordinator.
+func NewGCCoordinator(
+	executorInfos ExecutorInfoProvider,
+	jobInfos JobStatusProvider,
+	metaClient pkgOrm.ResourceClient,
+	gcNotifier GCNotifier,
+) *DefaultGCCoordinator {
+	return &DefaultGCCoordinator{
+		executorInfos: executorInfos,
+		jobInfos:      jobInfos,
+		metaClient:    metaClient,
+		notifier:      gcNotifier,
+	}
+}
+
+// Run runs the DefaultGCCoordinator.
+func (c *DefaultGCCoordinator) Run(ctx context.Context) error {
+	// We run a retry loop at the max frequency of once per second.
+	rl := ratelimit.New(1 /* once per second */)
+	for {
+		select {
+		case <-ctx.Done():
+			return errors.Trace(ctx.Err())
+		default:
+		}
+
+		jobReceiver, executorReceiver, err := c.initializeGC(ctx)
+		if err != nil {
+			log.L().Warn("GC error", zap.Error(err))
+			rl.Take()
+			continue
+		}
+
+		err = c.runGCEventLoop(ctx, jobReceiver.C, executorReceiver.C)
```

Comment:
What condition will this loop run again?
I scan the whole pr and find runGCEventLoop will not return if no error happen.

Reply:
You are right. This loop is for retrying on errors.

Reply:
So `Run` will be called multi-times?

Reply:
No, it will not.